doc/examples/segmentation/plot_trainable_segmentation.py

```diff
@@ -0,0 +1,168 @@
+    return list(itertools.chain.from_iterable(all_results))
+
+
+def compute_features(img, multichannel=True,
+                          intensity=True, edges=True, texture=True,
+                          sigma_min=0.5, sigma_max=16):
+    """Features for a single- or multi-channel image.
+    """
+    if img.ndim == 3 and multichannel:
+        all_results = (_compute_features_gray(
+                img[..., dim], intensity=intensity, edges=edges, texture=texture,
+                sigma_min=sigma_min, sigma_max=sigma_max) for dim in range(img.shape[-1]))
+        features = list(itertools.chain.from_iterable(all_results))
+    else:
+        features = _compute_features_gray(
+            img[..., dim], intensity=intensity, edges=edges, texture=texture,
+            sigma_min=sigma_min, sigma_max=sigma_max)
+    return np.array(features)
+
+
+def trainable_segmentation(img, mask, clf, multichannel=True,
+                           intensity=True, edges=True, texture=True,
+                           sigma_min=0.5, sigma_max=16, downsample=10):
+    """
+    Segmentation using labeled parts of the image and a random forest classifier.
+    """
+    print(img.shape)
+    t1 = time()
+    features = compute_features(im, multichannel=multichannel,
+                            intensity=intensity, edges=edges, texture=texture,
+                            sigma_min=sigma_min, sigma_max=sigma_max)
+    t2 = time()
+    training_data = features[:, mask > 0].T
+    training_labels = mask[mask>0].ravel()
+    data = features[:, mask == 0].T
+    t3 = time()
+    clf.fit(training_data[::downsample], training_labels[::downsample])
+    t4 = time()
+    labels = clf.predict(data)
+    t5 = time()
+    result = np.copy(mask)
+    result[mask == 0] = labels
+    print("compute features", t2 - t1)
+    print("fit", t4 - t3)
+    print("predict", t5 - t4)
+    return result
+
+
+from time import time
+# This image is in the public domain and could be added to the Gitlab data repo
+filename = 'https://upload.wikimedia.org/wikipedia/commons/thumb/b/b4/Normal_Epidermis_and_Dermis_with_Intradermal_Nevus_10x.JPG/1280px-Normal_Epidermis_and_Dermis_with_Intradermal_Nevus_10x.JPG'
+im = io.imread(filename)
+
+# Build a mask for training the segmentation.
+# Here we use rectangles but visualization libraries such as plotly
+# (and napari?) can be used to draw a mask on the image.
+mask = np.zeros(im.shape[:2], dtype=np.uint8)
+mask[:100] = 1
+mask[:170, :400] = 1
+mask[600:900, 200:650] = 2
+mask[330:430, 210:320] = 3
+mask[260:340, 60:170] = 4
+mask[150:200, 720:860] = 4
+
+sigma_min = 1
+sigma_max = 32
+clf = RandomForestClassifier(n_estimators=100, n_jobs=-1)
+result = trainable_segmentation(im, mask, clf, multichannel=True,
+                        intensity=True, edges=False, texture=True,
+                        sigma_min=sigma_min, sigma_max=sigma_max, downsample=15)
```

Comment:
This is just an example, but I'm thinking about adding this to the library. In this case, I would like to separate the training and the prediction steps, since typically one trains on a few images and then predicts on many more in batch.

So, something along the lines of:

```python
clf, predicted = train_segmenter(image, labels, clf, features=feature_func, ...)  # analogous to `fit_transform`
predictions = apply_segmenter(images, clf, features=feature_func, ...)  # analogous to `predict_proba`
```

There's an issue of keeping the feature transforms "in sync", but I don't have a good answer for that... =\

Reply:
Could we do `clf.features_func` inside `train_segmenter` so that we can just call `clf.features_func` inside `apply_segmenter`? I don't know if it's bad practice to do this... Adding an attribute to a class instance sounds pretty innocuous to me.

Reply:
Ha ha I spoke with Gael and this type of monkey-patching is definitely a bad practice :-), all the more than scikit-learn objects erase their attribute every time that the `clone` method is called. I'll think of something else, maybe something which is object-oriented.

Reply:
Perhaps he has encountered the problem of matching the feature computation to the classifier? Is there a proper way to use sklearn pipelines for this?